working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+```dart  
+class S implements Shareable {
+  int v = 0;
+}
+
+final s = S();
+
+await Isolate.run(() {
+  // Even though we are running in a different isolate we have
+  // access to the original `s` and mutations performed in this
+  // isolate will be visible to other isolates.
+  // 
+  // (Subject to memory model constraints).
+  s.v = 10; 
+});
+
+expect(equals(10), s.v);
+```
+
+To make state sharing between isolates simpler we also allow to declare static fields which are shared between all isolates in the isolate group. `@shared` global fields are required to have static type that is a subtype of `Shareable?`. 
+
+```dart 
+// All isolates within an isolate group share this variable.
+@shared int v = 0;
+```
+
+### Functions
+
+Given the parallel between functions and other objects it makes sense to declare that functions which close only over shareable state are themselves shareable.
+
+**Static type of a function literal which captures only `Shareable` variables is  `ShareableFunction<F extends Function>` which is subtype of both `F` and `Shareable`.**
+
+```dart
+void main() {
+  final Shareable s = ...;
+  final f = () => s;  // ShareableFunction<Shareable Function()>
+
+  final Object o;
+  final g = () => o;  // Not shareable
+}
+
+class S implements Shareable {
+  void Function() f; // compile time error
+  
+  ShareableFunction<void Function()> f;  // ok
+}
+```
+
+Shareable functions have one subtle behavior - they capture and share local variables, but not the global state of the isolate. Consider for example the following code:
+
+``` dart
+int global = 0; 
+
+void main() async {
+  final s = S();
+  
+  s.v = 42;
+  global = 42;
+  
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+
+    print(s.v);  // 42
+    s.v = 24;
+  });
+  
+  print(global);  // 42
+  print(s.v);  // 24
+}
+```
+
+This behavior might be somewhat surprising to developers - especially if a functions depends on or updates some mutable global state. Whether this is a problem in practice is unclear: sending a closures through a `SendPort` leads to similar behavior.
+
+### Records
+
+Records where all components are shareable are themselves shareable. 
+
+### Collections (`Iterable`, `List`, `Set`, `Map`)
+
+If `S` shareable then `Iterable<S>`, `List<S>` and `Set<S>` are shareable. If `SK` and `SV` are sharable then `Map<SK, SV>` is shareable. This imposes additional requirements on classes implementing these interfaces: whether some container is shareable depends on concrete type parameters of an instance. If class `C<T1, ..., Tn>` implements `List<F(T1, ..., Tn)>` then constructing an instance of `C<X1, ..., Xn>` , such that `F(X1, ..., Xn)` is shareable, should only be possible if `C<X1, ..., Xn>` satisfies restrictions imposed on shareable classes i.e. static type of each of its fields should be a subtype of `Shareable`. 
```

Comment:
So `List<T>` is shareable if `T` is shareable? How is this logic implemented? Are `List`s etc special cased in the CFE? Is there a way for users to define their own classes with this property, where the shareability of their class depends on the shareability of the type params?

Reply:
Since anyone can implement `List`, there is nothing preventing a `class MyList<T> implements List<T> { final _nonSharableObject = ...; ...}`, so my best guess is that it cannot possibly work unless the `List` *interface* somehow carries an "must be non-shareable except for `T`" constraint that all subtypes are forced to satisfy.

So, a conditional implements clause:
```dart
abstract mixin class List<E> implements Iterable<E>, Sendable if (E is Sendable) { ... }
```
That's ... unlikely to be useful for anything other than marker interfaces, so not a great language feature by itself.
But it would work. Doing `(<int>[42] as dynamic) is Sendable` would do the computation:
* runtime type is `List<int>`
* Does `List<int>` have `Sendable` as superinterface.
    * Search for superinterfaces which implements `Sendable`.
    * Find one conditional.
        * Check whether `int` implements `Sendable`.
        * It does, unconditionally.
        * Yes!
    * One found.
* Yes!


That also suggests a failure-case of:
```dart
class Foo<T extends Foo<T>> implements Sendable if (T is Sendable) {}
class Bar extends Foo<Bar> {}
```
is `Bar` sendable? Yes, if `Bar` is!
(Which is probably a "yes", because *conditionally* being sendable still means the class must be sendable in every place except where the type parameter occurs, there it can be only conditionally sendable.
And `Bar` doesn't have the type parameter anywhere, so it must just be unconditionally sendable. It might as well write `implements `Sendable`, which would be a valid subclass of something implementing `Sendable if ...`.)

So, a highly comlicated language feature, which doesn't really work for anything else than marker interfaces.
It's not going to be an easy sell.